lib/metadata.sh

```diff
@@ -4,75 +4,140 @@
 
-# Based on: https://github.com/heroku/heroku-buildpack-nodejs/blob/main/lib/metadata.sh
+# Legacy `key=value` format file used by older Python buildpack versions.
+LEGACY_BUILD_DATA_FILE="${CACHE_DIR:?}/build-data/python"
 
-source "${BUILDPACK_DIR:?}/lib/kvstore.sh"
+# Initializes the metadata store, preserving the file from the previous build if it exists.
+# Call this at the start of `bin/compile` before using any other functions from this file.
+function metadata::setup() {
+	if [[ -f "${METADATA_FILE}" ]]; then
+		# Rename the existing metrics file rather than overwriting it, so we can lookup values
+		# from the previous build (such as when determining whether to invalidate the cache).
+		mv "${METADATA_FILE}" "${PREVIOUS_METADATA_FILE}"
+	else
+		mkdir -p "$(dirname "${METADATA_FILE}")"
+	fi
 
-# Variables shared by this whole module
-BUILD_DATA_FILE=""
-PREVIOUS_BUILD_DATA_FILE=""
+	echo "{}" >"${METADATA_FILE}"
+}
 
-# Must be called before you can use any other methods
-meta_init() {
-	local cache_dir="${1}"
-	local buildpack_name="${2}"
-	BUILD_DATA_FILE="${cache_dir}/build-data/${buildpack_name}"
-	PREVIOUS_BUILD_DATA_FILE="${cache_dir}/build-data/${buildpack_name}-prev"
+# Retrieve the value of an entry in the metadata store for the current build.
+# Returns the empty string if the key wasn't found in the store.
+#
+# Usage:
+# ```
+# metadata::get "python_version"
+# ```
+function metadata::get() {
+	local key="${1}"
+	jq --raw-output ".${key} // empty" "${METADATA_FILE}"
 }
 
-# Moves the data from the last build into the correct place, and clears the store
-# This should be called after meta_init in bin/compile
-meta_setup() {
-	# if the file already exists because it's from the last build, save it
-	if [[ -f "${BUILD_DATA_FILE}" ]]; then
-		cp "${BUILD_DATA_FILE}" "${PREVIOUS_BUILD_DATA_FILE}"
-	fi
+# Retrieve the value of an entry in the metadata store from the previous successful build.
+# Returns the empty string if the key wasn't found in the store.
+#
+# Usage:
+# ```
+# metadata::get_previous "python_version"
+# ```
+function metadata::get_previous() {
+	local key="${1}"
 
-	kv_create "${BUILD_DATA_FILE}"
-	kv_clear "${BUILD_DATA_FILE}"
+	# Older versions of this buildpack used a `key=value` format file instead of JSON,
+	# so we need to support this file format/location too, so older caches can be read.
+	# We check for this file first, so that we correctly handle the case where an app
+	# downgraded and then re-upgraded buildpack version, so has both files in the cache.
+	if [[ -f "${LEGACY_BUILD_DATA_FILE}" ]]; then
+		# The legacy file contains one entry per line, of form `key=value`. Entries were written in an
+		# append-only manner so there could be duplicate entries for each key, so we return only the
+		# last matching entry in the file. The empty string is returned if the key wasn't found.
+		tac "${LEGACY_BUILD_DATA_FILE}" | { grep --perl-regexp --only-matching --max-count=1 "^${key}=\K.*$" || true; }
```

Comment:
(optional) Suggestion: Exercise this `get_previous` legacy logic under test. 

I poked around, and it looks like there's no unit test capability right now. You could add something like https://github.com/heroku/heroku-buildpack-ruby/blob/a6d124492aafe7f7ee0dd42d7a79284b6552d675/spec/unit/bash_functions_spec.rb#L94 that executes the code locally (but requires OS command parity). Or you could have an inline buildpack that writes some old data to that file, load the `metadata.sh`, and then `get_previous` and echo them (doesn't need to run the Python buildpack, just `metadata.sh` contents).

Reply:
(values like the previous uv version come from the metadata store, and are listed in the build output in the test assertion, proving it read the old value correctly: https://github.com/heroku/heroku-buildpack-python/blob/3c27de9acc7a7bbcaf27d281e1381cd81c793999/spec/hatchet/uv_spec.rb#L106)